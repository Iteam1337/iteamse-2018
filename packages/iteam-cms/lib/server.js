require('dotenv').config()
const cors = require('micro-cors')
const { ApolloServer } = require('apollo-server-micro')
const contentful = require('contentful')
const Query = require('./queries')
const typeDefs = require('./schema')
const getByContentType = require('./services/getByContentType')

const client = contentful.createClient({
  space: process.env.CONTENTFUL_SPACE,
  accessToken: process.env.CONTENTFUL_ACCESS_TOKEN,
})

const resolvers = {
  Query,
}

const server = new ApolloServer({
  playground: true,
  introspection: true,
  typeDefs,
  resolvers,
  context: () => ({
    client,
    getByContentType: getByContentType(client),
  }),
})

module.exports = cors({ origin: '*' })((req, res) => {
  if (req.method === 'OPTIONS') {
    res.end()
    return
  }
  return server.createHandler({ path: '/api/graphql' })(req, res)
})
