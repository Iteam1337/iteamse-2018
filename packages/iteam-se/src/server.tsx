import { InMemoryCache } from 'apollo-cache-inmemory'
import { ApolloClient } from 'apollo-client'
import { createHttpLink } from 'apollo-link-http'
import compression from 'compression'
import express from 'express'
import proxy from 'http-proxy-middleware'
import fetch from 'node-fetch'
import React from 'react'
import { ApolloProvider, renderToStringWithData } from 'react-apollo'
import { renderToStaticMarkup } from 'react-dom/server'
import { FilledContext, HelmetProvider } from 'react-helmet-async'
import { StaticRouter } from 'react-router-dom'
import serialize from 'serialize-javascript'
import App from './App'
import Html from './Html'
import { ServerStyleSheet, theme, ThemeProvider } from './theme'
import { redirectHelper } from './utils/serverHelpers'

interface StaticContext {
  statusCode?: number
}

const server = express()
const {
  RAZZLE_CMS_NODE_URL = '/api/graphql',
  RAZZLE_PUBLIC_DIR = '/api/graphql',
  RAZZLE_HOST = '/api/graphql',
} = process.env

server.use(compression())
server.use(
  '/api/graphql',
  proxy({
    pathRewrite: (path: string) => path.replace('/api/graphql', ''),
    target: RAZZLE_CMS_NODE_URL,
  })
)

server
  .disable('x-powered-by')
  .use(express.static(RAZZLE_PUBLIC_DIR!))
  .get(
    '/*',
    redirectHelper,
    async (req: express.Request, res: express.Response) => {
      try {
        const client = new ApolloClient({
          cache: new InMemoryCache(),
          link: createHttpLink({
            fetch,
            uri: RAZZLE_HOST,
          }),
          ssrMode: true,
        })

        const sheet = new ServerStyleSheet()

        const context: StaticContext = {}
        const helmetContext: FilledContext = {
          helmet: {},
        }

        const markup = sheet.collectStyles(
          <HelmetProvider context={helmetContext}>
            <ApolloProvider client={client}>
              <ThemeProvider theme={theme}>
                <StaticRouter context={context} location={req.url}>
                  <App />
                </StaticRouter>
              </ThemeProvider>
            </ApolloProvider>
          </HelmetProvider>
        )

        const content = await renderToStringWithData(markup)

        const apolloState = client.extract()
        const styleTags = sheet.getStyleElement()

        const html = (
          <Html
            apolloState={serialize(apolloState, { isJSON: true })}
            content={content}
            styleTags={styleTags}
          />
        )

        const staticMarkup = renderToStaticMarkup(html)

        if (context.statusCode) {
          res.status(context.statusCode)
        }

        const { helmet } = helmetContext
        const { meta, title } = helmet

        res.send(`
          <!doctype html>
            <head>
              ${title.toString()}
              ${meta.toString()}
            </head>
            ${staticMarkup}
        `)
      } catch (error) {
        console.error(error)

        res.status(500)
        res.end(
          `An error occurred. Please submit an issue to [https://github.com/apollographql/GitHunt-React] with the following stack trace:\n\n${
            error.stack
          }`
        )
      }
    }
  )

export default server
