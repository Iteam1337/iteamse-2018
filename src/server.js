import App from './App'
import React from 'react'
import { StaticRouter } from 'react-router-dom'
import express from 'express'
import compression from 'compression'
import { renderToStaticMarkup } from 'react-dom/server'
import { ApolloProvider, renderToStringWithData } from 'react-apollo'
import fetch from 'node-fetch'
import { ApolloClient } from 'apollo-client'
import { InMemoryCache } from 'apollo-cache-inmemory'
import { createHttpLink } from 'apollo-link-http'
import Html from './Html'
import serialize from 'serialize-javascript'
import { ThemeProvider, ServerStyleSheet } from 'styled-components'
import { theme } from './theme'

const server = express()

server.use(compression())

server
  .disable('x-powered-by')
  .use(express.static(process.env.RAZZLE_PUBLIC_DIR))
  .get('*', async (req, res) => {
    try {
      const client = new ApolloClient({
        cache: new InMemoryCache(),
        link: createHttpLink({
          fetch,
          uri: process.env.RAZZLE_CMS_NODE_URL,
        }),
        ssrMode: true,
      })

      const sheet = new ServerStyleSheet()
      const context = {}

      const markup = sheet.collectStyles(
        <ApolloProvider client={client}>
          <ThemeProvider theme={theme}>
            <StaticRouter context={context} location={req.url}>
              <App />
            </StaticRouter>
          </ThemeProvider>
        </ApolloProvider>
      )

      const content = await renderToStringWithData(markup)

      const apolloState = client.extract()
      const styleTags = sheet.getStyleElement()

      const html = (
        <Html
          apolloState={serialize(apolloState, { isJSON: true })}
          content={content}
          styleTags={styleTags}
        />
      )

      res.send(`<!doctype html>\n${renderToStaticMarkup(html)}`)
    } catch (error) {
      console.error(error)

      res.status(500)
      res.end(
        `An error occurred. Please submit an issue to [https://github.com/apollographql/GitHunt-React] with the following stack trace:\n\n${
          error.stack
        }`
      )
    }
  })

export default server
