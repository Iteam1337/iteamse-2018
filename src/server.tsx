import { InMemoryCache } from 'apollo-cache-inmemory'
import { ApolloClient } from 'apollo-client'
import { createHttpLink } from 'apollo-link-http'
import compression from 'compression'
import express from 'express'
import proxy from 'http-proxy-middleware'
import fetch from 'node-fetch'
import React from 'react'
import { ApolloProvider, renderToStringWithData } from 'react-apollo'
import { renderToStaticMarkup } from 'react-dom/server'
import { FilledContext, HelmetProvider } from 'react-helmet-async'
import { StaticRouter } from 'react-router-dom'
import serialize from 'serialize-javascript'
import App from './App'
import Html from './Html'
import { ServerStyleSheet, theme, ThemeProvider } from './theme'
import { redirectHelper } from './utils/serverHelpers'

const server = express()
const { RAZZLE_CMS_NODE_URL, RAZZLE_PUBLIC_DIR, RAZZLE_HOST } = process.env

server.use(compression())
server.use(
  '/cms',
  proxy({
    pathRewrite: (path: string) => path.replace('/cms', ''),
    target: RAZZLE_CMS_NODE_URL,
  })
)

server
  .disable('x-powered-by')
  .use(express.static(RAZZLE_PUBLIC_DIR!))
  .get(
    '/*',
    redirectHelper,
    async (req: express.Request, res: express.Response) => {
      try {
        const client = new ApolloClient({
          cache: new InMemoryCache(),
          link: createHttpLink({
            fetch,
            uri: RAZZLE_HOST,
          }),
          ssrMode: true,
        })

        const sheet = new ServerStyleSheet()
        const context = {}
        const helmetContext: FilledContext = {
          helmet: {},
        }

        const markup = sheet.collectStyles(
          <HelmetProvider context={helmetContext}>
            <ApolloProvider client={client}>
              <ThemeProvider theme={theme}>
                <StaticRouter context={context} location={req.url}>
                  <App />
                </StaticRouter>
              </ThemeProvider>
            </ApolloProvider>
          </HelmetProvider>
        )

        const content = await renderToStringWithData(markup)

        const apolloState = client.extract()
        const styleTags = sheet.getStyleElement()

        const html = (
          <Html
            apolloState={serialize(apolloState, { isJSON: true })}
            content={content}
            styleTags={styleTags}
          />
        )

        const staticMarkup = renderToStaticMarkup(html)

        const { helmet } = helmetContext
        const { meta, title } = helmet

        res.send(`
          <!doctype html>
            <head>
              ${title.toString()}
              ${meta.toString()}
            </head>
            ${staticMarkup}
        `)
      } catch (error) {
        console.error(error)

        res.status(500)
        res.end(
          `An error occurred. Please submit an issue to [https://github.com/apollographql/GitHunt-React] with the following stack trace:\n\n${
            error.stack
          }`
        )
      }
    }
  )

export default server
