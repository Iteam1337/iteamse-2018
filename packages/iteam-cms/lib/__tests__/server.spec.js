jest.mock('../services/getByContentType')
jest.mock('apollo-server-micro')

describe('#server', () => {
  let ApolloServer

  beforeEach(() => {
    ApolloServer = require('apollo-server-micro').ApolloServer
  })

  afterEach(() => {
    jest.clearAllMocks()
    jest.resetModules()
  })

  it('should setup server', () => {
    require('../server')

    expect(ApolloServer).toMatchSnapshot()
  })

  it('should setup context for GraphQL', () => {
    require('../server')

    expect(ApolloServer.mock.calls[0][0].context()).toMatchSnapshot()
  })
})
