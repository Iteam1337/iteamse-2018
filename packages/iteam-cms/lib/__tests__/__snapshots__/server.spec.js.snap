// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#server should setup context for GraphQL 1`] = `
Object {
  "client": undefined,
  "getByContentType": [MockFunction getByContentType],
}
`;
